### PR TITLE
Issue # 708 Draggable List widget for Touch UI Components

### PR DESCRIPTION
**Pull request by kaushalmall**
screen shot 2016-05-08 at 2 47 22 pm

### CODE REVIEW

**Comment by kaushalmall** [comment]
## Purpose
The Draggable List component lets you create an ordered list of elements that are generated by a datasource. The ACS Query Autocomplete datasource can be used with it as well.

## How to Use

*Set up your component and its Touch UI dialog as usual.
*Add a new widget node with `sling:resourceType = acs-commons/components/authoring/draggablelist`
*If using the ACS Query Autocomplete datasource, set the `predicates` and the `displayProperty` properties
*Create a new `datasource` node under the draggablelist widget node.
*If using the ACS Query Autocomplete datasoruce, set `sling:resourceType = acs-commons/granite/ui/components/form/queryautocomplete/datasource`
*If using your own datasource, set the `sling:resourceType` accordingly.

## Sample Datasource.

{% highlight jsp %}
<%@include file="/libs/granite/ui/global.jsp" %>
<%@page session="false"
        import="com.adobe.cq.commerce.common.ValueMapDecorator,
                com.adobe.granite.ui.components.ds.DataSource,
                com.adobe.granite.ui.components.ds.SimpleDataSource,
                com.adobe.granite.ui.components.ds.ValueMapResource,
                org.apache.sling.api.resource.ResourceMetadata,
                java.util.ArrayList,
                java.util.HashMap,
                java.util.List" %>
<%
    List<Resource> resourceList = new ArrayList();

    for( int i = 0; i <= 3; i ++) {
        ValueMap vm = new ValueMapDecorator(new HashMap<String, Object>());
        vm.put("value", i);
        vm.put("text", i);
        resourceList.add(new ValueMapResource(resourceResolver, new ResourceMetadata(), "nt:unstructured", vm));
    }

    DataSource ds = new SimpleDataSource(resourceList.iterator());
    request.setAttribute(DataSource.class.getName(), ds);
%>

{% endhighlight %}

## Example Dialog

{% highlight xml %}
<?xml version="1.0" encoding="UTF-8"?>
   <jcr:root xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:nt="http://www.jcp.org/jcr/nt/1.0" xmlns:sling="http://sling.apache.org/jcr/sling/1.0" jcr:primaryType="nt:unstructured" jcr:title="List" sling:resourceType="cq/gui/components/authoring/dialog" helpPath="en/cq/current/wcm/default_components.html#List">
      <content
      	jcr:primaryType="nt:unstructured"
      	sling:resourceType="granite/ui/components/foundation/container">
      	<layout
      		jcr:primaryType="nt:unstructured"
      		margin="false"
      		sling:resourceType="granite/ui/components/foundation/layouts/fixedcolumns"/>
      	<items
      		jcr:primaryType="nt:unstructured">
      		<column
      			jcr:primaryType="nt:unstructured" sling:resourceType="granite/ui/components/foundation/container">
      			<items
      				jcr:primaryType="nt:unstructured">
      				<list
      					jcr:primaryType="nt:unstructured"
      					displayProperty="jcr:content/jcr:title"
      					fieldLabel="Draggable List"
      					name="./list"
      					sling:resourceType="acs-commons/components/authoring/draggablelist">
      					<datasource
      						jcr:primaryType="nt:unstructured"
      						sling:resourceType="mytestcomponents/datasources/draggablelist"/>
      				</list>
      			</items>
      		</column>
      	</items>
      </content>
   </jcr:root>
{% endhighlight %}

**Comment by kaushalmall** [comment]
@schoudry you mind looking at this real quick? Make sure I didn't miss anything or anyways I can make this better?

**Comment by schoudry** [comment]
@kaushalmall looks good to me, couple of things...

1) draggable css class i guess is too generic, change it to acs-commons.draggable-list ?

2) follow the same convention as data-init ( `CUI.DraggableList.init($("[data-init~=draggable-list]"));` ) and change filenames - draggablelist to draggable-list ?

**Comment by kaushalmall** [comment]
@schoudry I did a few more changes in addition to what you had suggested. Please have a look.

**Comment by schoudry** [comment]
@kaushalmall looks good, cool widget

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/draggablelist.css [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+ul.acs-commons-draggablelist-draggable li .close {
+    position: absolute;
+    right: 0;
+    top: 0;
+    padding: 0.125rem;
+    margin: 0.6875rem;
+}
+
+/*.dropzone {
```
Comment (davidjgonzalez) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/draggablelist.css]: Can we nix these commented styles?

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/draggablelist.js [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+
+    CUI.CopyDragAction = new Class({
+        extend: CUI.DragAction,
+        construct: function (event, source, dragElement, dropZones, restrictAxis) {
+            this.clone = dragElement.clone();
+            this.clone.hide();
+            this.clone.removeClass("is-dragging");
+            //this.clone.addClass("dragging");
```
Comment (davidjgonzalez) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/draggablelist.js]: Can these commented lines be removed?

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/draggablelist.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        },
+
+        dropZone: null,
+
+        dragStart: function (event) {
+            if ($(event.target).hasClass("close")) return; // Don't drag on close button!
+            if ($(event.target).hasClass("scf-js-item-action")) return; // Don't drag on close button!        
+            event.preventDefault();
+
+            //this.dragging = $(event.target).closest("li");
```
Comment (davidjgonzalez) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/draggablelist.js]: Same - can we remove commented out code?

**Comment by kaushalmall** [comment]
@davidjgonzalez done. thanks.

**Pull request merged**